Replace "loose" options with a top-level "assumptions" object (#5)
babel/rfcs: rfcs/0000-top-level-assumptions.md

# RFC: Loose

- Repo: [`babel/babel`](https://github.com/babel/babel)
- Start Date: 2020-06-08
- RFC PR: <!-- leave this empty, to be filled in later -->
- Related Issues: <!-- if relevant -->
- Authors: Nicolò Ribaudo
- Champion: Nicolò Ribaudo
- Implementors: <!-- the names of everyone who will work on the PR. you can leave this empty if you would like someone else to work on it -->

# Summary

Many of our plugins have a `loose` option, which enables the compiler to make some assumptions about the code you are writing. It uses these to ignore certain edge cases and generate a smaller or faster output.

The `loose` option has been around since `6to5`: at first as a top level option, and then it was split into each plugin as a catch all for anything that we wanted to be less spec compliant.

> However, `loose` has some problems: namely that the term itself is not clear what it means, and _should_ be more related to a part of the language rather than to the plugin that it's compiling it (see the "Motivation" section for more details). A lot of this behavior is documented with a single example, so in the future we’ll want to be specific with the assumptions themselves and maybe provide a set of tests/codesandboxes explaining differences.

Reviewer: ```suggestion
However, `loose` has some problems: namely the term itself is not really descriptive at all and _should_ be named after the part of the language it affects rather than the plugin that's compiling it (see the "Motivation" section for more details). A lot of this behavior is documented with a single example, so in the future we’ll want to be specific with the assumptions themselves and maybe provide a set of tests/codesandboxes explaining the differences.
```

> This RFC proposes introducing a new *top-level* option, `"assumptions"`, which is an object containing different flags which can simplify the output code.

Reviewer: ```suggestion
This RFC proposes introducing a new *top-level* option, `"assumptions"`, which is an object containing different flags that can simplify the code that Babel generates.
```

# Basic example

```jsonc
// babel.config.json
{
  "assumptions": {
    "noDocumentAll": true,
    "noGetterSideEffects": true,
    "iterableIsArray": true
  },
  "presets": ["@babel/preset-env"]
}
```

<details>
<summary>Input / Output</summary>

```javascript 
// input code

for (const el of arr) {
  el.logger?.log("Ok!");
}
```

```javascript 
// output code

for (let i = 0; i < arr.length; i++) {
  const el = arr[i];
  el.logger == null ? void 0 : el.logger.log("Ok!");
}
```
</details>
<br />

```jsonc
// babel.config.json
{
  "assumptions": {
    "noDocumentAll": false,
    "noGetterSideEffects": true,
    "iterableIsArray": true
  },
  "presets": ["@babel/preset-env"]
}
```

<details>
<summary>Input / Output</summary>

```javascript 
// input code

for (const el of arr) {
  el.logger?.log("Ok!");
}
```

```javascript 
// output code

for (let i = 0; i < arr.length; i++) {
  const el = arr[i];
  el.logger === null || el.logger === void 0 ? void 0 : el.logger.log("Ok!");
}
```
</details>

# Motivation

The `loose` options have different problems:

> - _"What is loose?"_ It's not always clear what assumptions `loose` makes. You always have to check in the docs, end even the docs often only partially mention what `loose` means in that specific context.

Reviewer: ```suggestion
- _"What is loose?"_ Since `loose` is not descriptive, it requires users to check a plugin's docs to see what assumptions it makes (and the docs often only partially mention what `loose` means in a specific context).
```

- Some plugins should be aware of other plugins' `loose` option value. For example, we sometimes need to partially transpile optional chaining in the class properties plugin ([babel/babel#11248](https://github.com/babel/babel/pull/11248)). Should we add a new `optionalChainingLoose` option to `proposal-private-methods` and `proposal-class-properties`? Or should we use a cross-plugin communication channel to share it? What about when someone has the private methods plugin enabled, but not the optional chaining one?

This RFC also solves another problem, which is not directly related to the points above:
- `@babel/preset-env` has a `loose` option, which is forwarded to all the plugins it enables. When a user wants to set `loose` only for a specific feature, they have to manually install the plugin and explicitly add it to their configuration ([babel/babel#6978](https://github.com/babel/babel/issues/6978)).

# Detailed design

## New configuration option

This RFC introduces a new top-level option: `assumptions`. This is an object containing different flags to mark different assumptions about the input code as safe and thus allow optimizing them.

```typescript
type Assumptions = {

>   [assumption: string]: boolean

Author: My plan for this is to default to `true` for compat in Babel 7, and default to `false` in Babel 8 (since currently it's the only thing where spec compliancy is opt-in)

>   [assumption: string]: boolean

Author: Yeah I described them as simple booleans exactly to avoid thinking about merging them :sweat_smile: 

I feel like we can just use longer names instead of nesting objects.

>   [assumption: string]: boolean

Reviewer 2: We can also consider deprecate `spec` of `preset-env`/`transform-arrow-functions` and encourage to use `assumptions`, so we can eventually remove `spec` in Babel 8.

};
```

> This new options should be allowed in the following locations:

Reviewer: ```suggestion
This new option should be allowed in the following locations:
```

- In `overrides` blocks. This is useful, for example, to enable a possible `noDocumentAll` assumption on a folder containing server-side code and not inside one containing client-side code.
- In `env` blocks. This is mostly included for completeness, but I can't think about an use case for it: assumptions are about the _input code_, which isn't affected by env variables.

Additionally, `"assumptions"` should also be allowed inside presets.
There are two main kinds of presets: framework-specific presets, like `babel-preset-react-app`, and company-wide presets, used to share the same plugins accross different projects. Since assumptions describe the input code and not the used plugins, this new option fits better in company-wide presets where the person writing the preset and the person writing the input code have less "degrees of separation", but there are also use cases in framework-specific presets: for example, `babel-preset-react-app` [enables](https://github.com/facebook/create-react-app/blob/c87ab79559e98a5dae2cd0b02477c38ff6113e6a/packages/babel-preset-react-app/create.js#L151-L154) `loose: true` for `@babel/plugin-proposal-class-properties`.

## Configuration merging

> Multiple `"assumptions"` object should be merged using `Object.assign`, instead of by overwriting the object as it happens for other options.

Reviewer: ```suggestion
Multiple `"assumptions"` objects should be merged using `Object.assign`, instead of by overwriting the object as it happens for other options.
```

> This makes it easy, for example, to enable one additional assumption for a specific folder. Also, it still keeps the ability of disabling an assumption simply by setting it to `false`.

Author: But what it `preset-react` is included in `preset-pedantic`? `preset-pendantic` might just want to expose `preset-react`'s behavior but override it's assumptions.

I think we can handle it as we do with the other boolean options (`retainLines`, `comments`) or non-boolean options (`sourceType`, ...), which is that the last one wins.

> This makes it easy, for example, to enable one additional assumption for a specific folder. Also, it still keeps the ability of disabling an assumption simply by setting it to `false`.

Reviewer 2: I am not sure if `retainLines` and `comments` is as popular as `loose` in presets. We should somehow remind preset authors to choose `assumptions` wisely: It may be pretty hard for end users to read through a long list of assumptions.

> This makes it easy, for example, to enable one additional assumption for a specific folder. Also, it still keeps the ability of disabling an assumption simply by setting it to `false`.

Author: I would be ok with disallowing `: false` in presets initially, and wait for feedback.

> This makes it easy, for example, to enable one additional assumption for a specific folder. Also, it still keeps the ability of disabling an assumption simply by setting it to `false`.

Author: Added to the RFC text

2. With the current implementation, `skipForOfIterationClosing` and `arrayIndexedIteration` should be merged in a single option.
3. Currently, the `loose` option must be the same for these plugins or it will throw an error. We have a workaround in `@babel/preset-env` to allow setting `loose` there differently from these plugins ([babel/babel#11634](https://github.com/babel/babel/pull/11634))
4. `?.` also needs to be compiled by the private fields and methods plugins ([babel/babel#11248](https://github.com/babel/babel/pull/11248), but currently there is no way to set it to `loose`
5. The current default behavior is to consider this assumption as valid, and only produce 100% spec-compliant code when the `spec` option is enabled.

# Drawbacks

- This RFC introduces a very big number of new top-level options (19), and makes it relatively cheap to add new ones. Having a big number of options can add more "tooling fatigue" on the shoulders of our users, and it can make it hard for us to properly document them. However, these new options replace 22 existing plugin options, many of which have the same name (`loose`) but different behaviors.

> - ([@JLHwung](https://github.com/JLHwung)) Some `assumptions` are only effective in one plugin, but users will have to go through their presets/plugins to see if they have opt-in to this plugins. Let's say there is an assumption in a stage-1 proposal and the users does not know it is no-op for their config, it may inadvertently turn into a long list of `assumptions` and the no-op assumptions can kick in once users upgrades `preset-env` which may includes it once its stage advances.

Reviewer 2: On possible assumptions in early stage plugins, can we limit `assumptions` to be a something that related to stage 3 features or standard spec only? So we don't add assumptions about stage-1 plugins until they are mature enough. The assumptions can still be a valid plugin-specific optionl.

Author: I 99%* agree! Maybe we should have the same policy we have for `shippedProposals`?

\* only 99% because, for example, the `#foo in obj` transform is obviously influenced by the assumptions we make when transforming private fields (`privateFieldsAsPublic`).

Reviewer 2: In your case `privateFieldsAsPublic` falls to _shipped_ feature assumptions. A stage-1 proposal implementation can choose to respect such assumptions, but it can not, i.e. register a global `privateInNeverNested` assumption, until it is mature enough -- So other plugins might have to take care of that.

Author: 100% agree now :smile: 

I'll add a "Policy" section.

    ```

I prefer the first alternative, since it makes it possible for our users to gradually migrate to the new options and gradually makes the `loose` options noops in users' configs while migrating.

# How we teach this

This RFC proposed the addition of a new set of options in a way that every option can affect multiple plugins, and a plugin can be affected by multiple options.

Documenting the different `loose` and `loose`-like options wasn't too hard: every plugin had a bunch of those options, and we could document them in the plugin's page.

A possible structure for the documentation could be a two-entry table, with all the plugins in the rows and the assumptions in the columns, marking which assumption affects which plugins. However, this would lead to an enormous highly sparse table, impossible to fit in a normal web page.

I think that the best way of documenting these options is to put all of them in a single long list, divided in two sections: assumptions that affect standard features, and assumptions that only affect proposals. For each assumption we should write:
- The option name
- It's description
- An example (which could be interactive, using CodeSandbox)

> - The affected plugins

Reviewer 2: Maybe worthy to offer a green engines list for each assumptions. Like if you are _not_ transpiling down to these browsers, this assumption is a no-op for you. i.e. People should not worry about what is `setSpreadProperties` if they are not traspiling object spread anyway.

Author: We could even automatically generate it by computing the union of all the affected plugins' targets

In the docs of each affected plugin, we should write the assumptions that affect it and link to the corresponding entry in the assumptions list. We shouldn't duplicate the description of what each assumption does, because duplicating it would likely lead to the two copies become out of sync.

We can also provide a tool to check that the all the assumptions hold. This can't be done with an ESLint plugin because almost all the assumptions are about runtime semantics hard to statically analyze, but we can do it using a Babel plugin which should run while testing, and which injects assertions in the transpiled code:

```javascript 
// input code

foo?.bar;
```

```javascript
// output code

(() => {
  const _tmp = foo;
  if (_tmp == null && _tmp !== null && _tmp !== undefined) {
    throw new Error(`
      Invalid assumption - noDocumentAll.
      
      "noDocumentAll: true" implies that ?. and ?? are never
      used with the special document.all object.
      This assumption has been violated, so you should
      disable it in your config.
    `);
  }

>   reutrn _tmp

Reviewer 3: ```suggestion
  return _tmp
```

})?.bar
```

<details>
<summary>Config</summary>

```jsonc
// babel.config.json
{
  "assumptions": {
    "noDocumentAll": true
  },
  "env": {
    "test": {
      "plugins": ["@babel/validate-assumptions"]
    }
  }
}
```

</details>

# Open questions

- Should we try to pass to the presets at least a _partial_ `assumptions` object? Currently `@babel/preset-env` relies on `loose` to enable/disable the `typeof-symbol` plugin.

> - Should we validate the list of `assumptions` in `@babel/core`, and disallow unknown ones? This would make it impossible for third-party plugins to introduce their own assumptions, but it also means that it's easier for us to introduce new assumptions without risking ecosystem incompatibilities.

Reviewer: My gut reaction is to validate `assumptions`.

However, `loose` has some problems: namely the term itself is not really descriptive at all and _should_ be named after the part of the language it affects rather than the plugin that's compiling it (see the "Motivation" section for more details). A lot of this behavior is documented with a single example, so in the future we’ll want to be specific with the assumptions themselves and maybe provide a set of tests/codesandboxes explaining the differences.

This RFC proposes introducing a new *top-level* option, `"assumptions"`, which is an object containing different flags that can simplify the code that Babel generates.
These flags have two characteristics:
- They specify _something_ about a specific language feature, and not about a specific Babel plugin.
- Different flags can toggle different unrelated optimizations in the same plugin, instead of toggling both of them with a single generic `"loose"` option.

# Basic example

```jsonc
// babel.config.json
{
  "assumptions": {
    "noDocumentAll": true,

>     "noGetterSideEffects": true,

Reviewer 2: This assumption is not documented in the assumption lists. It is used in `optional-chaining` which assumes `a.b` is not an accessor.

Author: Thanks for catching this!

- preset &lt; `babel.config.json` &lt; `.babelrc.json` (far from the compiled file)&lt; `.babelrc.json` (near to the compiled file) &lt; programmatic options
- top-level &lt; `env` section &lt; `overrides` &lt; `overrides.env`

Assumptions set inside a preset are not boxed inside the preset but are applied to all the plugins, for two reasons:
1. conceptually, the assumptions describe the input file and not a specific plugin enabled inside the preset;
2. practically, this makes it possible to have a single "personal" or "company" preset containing assumptions which apply to a user's coding practices, and re-use it in different projects without duplicating the list of assumptions.

In order to expose assumptions defined inside presets to every plugin, we need to first resolve and instantiate all the presets, and then all the plugins with the resolved assumptions. This also means that we cannot provide the defined assumptions list to the presets, because it hasn't been finalized yet when they are instantiated.
This is implemented by [babel/babel#11689](https://github.com/babel/babel/pull/11689)).

## Plugin API

> The first parameter passed to the plugins (often known as `api`) should have a new method: `assumptions(name: string): boolean`, which returns whether or not the assumption has been enabled. If a plugin is asking about an assumption not supported by the used `@babel/core` version, it will return `false`.

Author: ```suggestion
The first parameter passed to the plugins (often known as `api`) should have a new method: `assumptions(name: string): boolean | undefined`, which returns whether or not the assumption has been enabled. If a plugin is asking about an assumption not enabled or not supported by the used `@babel/core` version, it will return `undefined`.
```

(idea from @jridgewell: https://github.com/babel/babel/pull/12407#discussion_r536511944)

By returning `undefined` for assumptions not explicitly set to `false`, plugins can customize their default behavior:
1. They can use `options.loose` as the default value of the assumption they are using.
1. Currently we are hard-coding `newableArrowFunctions` to `true` in `@babel/core` for backward compatibility, but we could move the default value to the plugin itself.

In general, I don't like making assumptions default to `true` in plugins (all assumptions should default to `false`, so that users don't have to also learn about the default values), but I think that if there is an explicit option it's ok.

What do you think?

This is implemented as a function and not as an object whose properties reflect the assumptions because it configures the plugins' caching (they are reinstantiated when an assumption they use changes).

## Assumptions list

The different `loose` or `loose`-like options we currently have in our plugins map to these assumptions:

> ℹ️ I'm not particularly attatched to these option names: I tried to choose something descriptive for their behavior, but they can all change.

| Assumption | Behavior | Current option | Repl | Notes |
|:-----------|:---------|:---------------|:----:|:-----:|
| `ignoreToPrimitiveHint` | Transform `` `a${x}b` `` to `"a" + x + "b"` instead of `"a".concat(x, "b")` | `loose` in `transform-template-literals` |

> | `mutableTemplateObject` | Don't use `Object.freeze` for the template object created for tagged template literals. This effectively means using the `taggedTemplateLiteralLoose` helper instead of `taggedTemplateLiteral` | `loose` in `transform-template-literals` |

Author: `noFrozenMutationAttempts`?

| `ignoreFunctionLength` | The `.length` of a function should be cropped at the first default argument. When this option is enabled, ignore this spec requirement and don't rely on the `arguments` object | `loose` in `transform-parameters` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=GYVwdgxgLglg9mABMMAKAhgGkQI0QXkQCZsJsATAxAZgEpEBvAKAF8g&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |

> | `iterableIsArray` | When using an iterable (in array destructuring, for-of or with spreads), assume that the iterable object is an `Array` | `loose` in `transform-destructuring` and `transform-spread`, `assumeArray` in `transform-for-of` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBCELwQ2gOmQIwLoG4BQpLwB4A0EyiAnmrFNkA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) (it doesn't support `assumeArray`) | 1 |

Author: `arrayLikeIsIterable` makes things like `{ 0: "a", 1: "b", length: 2 }` iterable, in addition to all the normal iterables. This is especially useful when working, for example, with DOM collections that _should_ be iterable but aren't in all the browsers, or with libraries that implement `Symbol.iterator` only when `Symbol` exists.

`iterableIsArray` is the opposite: rather than extending the definition of iterable, it restricts it. It ignores the iterator protocol, making it possible to produce much smaller code. For example, `for (let x of arr) {}` can be compiled to `for (let i = 0; i < arr.length; i++) { let x = arr[i] }`.

So the first one violates the spec by allowing more (for better compatibility), while the second one violates the spec by allowing less (for smaller generated code).

> | `iterableIsArray` | When using an iterable (in array destructuring, for-of or with spreads), assume that the iterable object is an `Array` | `loose` in `transform-destructuring` and `transform-spread`, `assumeArray` in `transform-for-of` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBCELwQ2gOmQIwLoG4BQpLwB4A0EyiAnmrFNkA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) (it doesn't support `assumeArray`) | 1 |

Author: Yes, it does.

| `arrayLikeIsIterable` | Allow array-like objects to be used where an iterable is expected. This can be useful, for example, to iterate DOM collections in older browsers | `allowArrayLike` in `transform-destructuring`, `transform-spread` and `transform-for-of` | | 1 |

> | `skipForOfIteratorClosing` | When using `for-of` with an iterator, it should always be closed with `.return()` and with `.throw()` in case of an error. This option allows skipping those methods | `loose` in `transform-for-of` | | |

Author: With `arrayLikeIsIterable` we still call `.throw()`/`.return()`, because we still respect the iterator protocol when it's there.
`iterableIsArray` will not be _that common_, because (for exmaple) it doesn't support `Map` and `Set`.

> | `objectRestNoSymbols` | When using rest in object destructuring, don't copy symbol keys. This effectively means using the `objectWithoutPropertiesLoose` helper instead of `objectWithoutProperties` | `loose` in `transform-destructuring` and `proposal-object-rest-spread` _(not documented)_ | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBA3hIDQQHTIEYQL4QLwQMYG4g&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |

Author: What other could we simplify with this?

> | `objectRestNoSymbols` | When using rest in object destructuring, don't copy symbol keys. This effectively means using the `objectWithoutPropertiesLoose` helper instead of `objectWithoutProperties` | `loose` in `transform-destructuring` and `proposal-object-rest-spread` _(not documented)_ | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBA3hIDQQHTIEYQL4QLwQMYG4g&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |

Reviewer 4: - `defineEnumerableProperties`
- `objectSpread`
- `objectWithoutProperties`
- Everything that uses `Object.assign` and doesn't check for `getOwnPropertySymbols`
- `instanceof`
- `for await`
- `ignoreToPrimitiveHint`

> | `objectRestNoSymbols` | When using rest in object destructuring, don't copy symbol keys. This effectively means using the `objectWithoutPropertiesLoose` helper instead of `objectWithoutProperties` | `loose` in `transform-destructuring` and `proposal-object-rest-spread` _(not documented)_ | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBA3hIDQQHTIEYQL4QLwQMYG4g&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |

Reviewer 2: > defineEnumerableProperties
objectSpread
objectWithoutProperties
...

That's really good point! I lean to introduce specific assumptions for each of them. Although a generic assumption ("I don't use symbols") can be handy, it is actually an umbrella term as we may add more optimization assumptions on "I don't use symbols", which could slip into the case of `"loose"` option that we try to avoid in this RFC. Generally the more an assumption is specific, the more easier people can reason about how this assumption might impact the transpiled output.

We can add tags/groups to the assumptions docs instead - So people can filter a list of assumptions under the term "I don't use Symbol".

| `setComputedProperties` | When using computed object properties, use `[[Set]]` semantics (i.e. use an assignment) instead of `[[Define]]` | `loose` in `computed-properties` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBCELwQ3gKAhA2gDwLoC4ICYAaRAXyA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |
| `setClassMethods` | When declaring classes, use `[[Set]]` semantics (i.e. use an assignment) instead of `[[Define]]`. This doesn't preserve the correct enumerability. | `loose` in `transform-classes` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=true&code_lz=MYGwhgzhAECC0G8BQ1oDMD2GAUBKRAvkgUA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |

> | `setPublicClassFields` | When using computed public class fields, use `[[Set]]` semantics (i.e. use an assignment) instead of `[[Define]]` | `loose` in `proposal-class-properties` | [🔗](https://babeljs.io/repl#?browsers=chrome%2070&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAECC0G8BQ1oDMCWBTEATaAvNAIwDcSAvkA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |

Author: @existentialism @JLHwung @hzoo Thoughts?

> | `setPublicClassFields` | When using computed public class fields, use `[[Set]]` semantics (i.e. use an assignment) instead of `[[Define]]` | `loose` in `proposal-class-properties` | [🔗](https://babeljs.io/repl#?browsers=chrome%2070&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAECC0G8BQ1oDMCWBTEATaAvNAIwDcSAvkA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |

Reviewer 2: I lean to specific assumptions unless we change the interface of assumptions from `{ [string]: boolean }` to `{ [string]: object }`. Assumptions is an advanced feature so I think we should prefer precision over ergonomics, as an advanced user can always add dynamics to assumptions:

```js
const useSetInsteadOfDefine = true;
const assumptions = {
  setPublicClassFields: useSetInsteadOfDefine
}
```

> | `setPublicClassFields` | When using computed public class fields, use `[[Set]]` semantics (i.e. use an assignment) instead of `[[Define]]` | `loose` in `proposal-class-properties` | [🔗](https://babeljs.io/repl#?browsers=chrome%2070&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAECC0G8BQ1oDMCWBTEATaAvNAIwDcSAvkA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |

Author: I really don't want to use something like `{ [string]: object }` since it makes it harder to reason about options merging.

Whenever we feel like namespacing assumptions (e.g. `{ useSet: { methods: true, fields: true } }`, I would advocated for flattening that object.

| `privateFieldsAsProperties` | Instead of storing private fields and methods using a `WeakMap` or a `WeakSet`, define them as own non-enumerable properties of the class instance | `loose` in `proposal-class-properties` _(not documented)_, in `proposal-private-methods` and `proposal-private-property-in-object` | [🔗](https://babeljs.io/repl#?browsers=chrome%2070&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAECC0G8BQ1oGIBmBLApiAJtALzQCMA3CtFWgLY4AuAFgPb4AUAlIlQJABOjAK78AdtGZYIAOky4ClVAF8kSoA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) | 2 |
| `superAsFunctionCall` | When this option is enabled, `super(arg1)` will be transpiled to `BaseClass.call(this, arg1)`. This means that it won't work with native classes or with built-ins, but only with compiled classes or ES5 constructors | `loose` in `transform-classes` | |

> | `constantSuper` | The `super` binding in classes can be changed using `setPrototypeOf`, so it's not possible to statically know it. With this option Babel can assume that the superclass is never changed at runtime. | `loose` in `transform-classes`, `proposal-class-properties`, `proposal-private-methods` and `proposal-decorators`. It should also be added to `transform-object-super`. | [🔗](https://babeljs.io/repl#?browsers=chrome%2040&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEBC0G8BQ1oFsCmAXAFgewBMAKASkWjBAwCcsiAiWAQnpIG5oBfJbpUSGAGFEKdNnzEyCClVoNBLdlx5I-4KNACC0DAA8sGAHYEY8ZKmoBXQ6RGpUAegfQAKgAkAkgGVoAGQ8AcgCiTKKoEJYADjQAdJi4hKRsoty8APIARgBWGMBYMRDYAArUeFhlAJ7RaQBmRJoxkaXlWFUYADTQgo3NldHsqoYYAO5apDFWNuxAA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env&prettier=false&targets=&version=7.10.1&externalPlugins=) |

Author: While working on #12497 (assumptions for class fields), I realized that I didn't notice this `loose` behavior when writing this RFC (https://github.com/babel/babel/pull/12497#issuecomment-744636617).

I have added it here, but I'm open to other names (`staticSuper`, `immutableSuper`, `noSuperReassign`, `constantSuperClass`, `constantExtends`, ...). My rationale for `constant` is that the behavior is similar to `const` variables: the _value_ of the super class can be mutated, but not the super class itself (i.e. the binding).

Reviewer 2: 👍 for `constantSuper`.

> | `inheritAsObjectCreate` | Use the `inheritsLoose` helper instead of `inherits` for derived classes | `loose` in `transform-classes` | |

Reviewer 3: The difference between the two is that `inheritsLoose` makes <code><var>subClass</var>.prototype.constructor</code> enumerable and doesn’t set <code><var>subClass</var>.[[Prototype]]</code> to <code><var>superClass</var></code>.

> | `inheritAsObjectCreate` | Use the `inheritsLoose` helper instead of `inherits` for derived classes | `loose` in `transform-classes` | |

Author: It's really hard to choose names for these options, so I'd be happy to rename it if you have any suggestion.

> | `inheritAsObjectCreate` | Use the `inheritsLoose` helper instead of `inherits` for derived classes | `loose` in `transform-classes` | |

Reviewer 4: Loose does `__proto__ = superClass`. I think the differences are

- strict supports `null` superclass
- strict uses define for `constructor`, loose uses set.
- strict will try to use `Object.setPrototypeOf` but falls back to `__proto__` setting

Honestly, I think we should just drop loose helper.

> | `inheritAsObjectCreate` | Use the `inheritsLoose` helper instead of `inherits` for derived classes | `loose` in `transform-classes` | |

Author: We could not implement this for now, and defer the decision if users ask for it

| `noDocumentAll` | When compiling the `??` and `?.` operators, assume that they are never used with `document.all` and thus `== null` is safe | `loose` in `proposal-optional-chaining` and `proposal-nullish-coalescing-operator` | [🔗](https://babeljs.io/repl#?browsers=chrome%2070&build=&builtIns=false&spec=false&loose=false&code_lz=IYfgdARg3AUMAEIT2kA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) | 4 |
| `setModuleMeta` | When compiling ESM to CJS, Babel defined a non-enumerable, non-witable, non-configurable `__esModule` property on the `exports` object. When this option is enabled, that property is set using a simple assignment | `loose` in `transform-modules-commonjs` |
| `copyReexports` | When re-exporting an imported value, assume that it's value doesn't change and set it with a simple assignment | `loose` in `transform-modules-commonjs`, `transform-modules-umd`, `transform-modules-amd` _(not documented)_ | | |

> | `newableArrowFunctions` | Do not check that arrow functions are never instantiated using `new` | `spec` in `transform-arrow-functions` | | 5 |

Author: In the future we might also want to introduce this for object/class methods, since they also shouldn't be newable according to the spec.

We could either name this `newableFunctions`, or decide to keep them separate (`newableArrowFunctions` and `newableMethods`).

> | `setModuleMeta` | When compiling ESM to CJS, Babel defined a non-enumerable, non-witable, non-configurable `__esModule` property on the `exports` object. When this option is enabled, that property is set using a simple assignment | `loose` in `transform-modules-commonjs` |

Reviewer 3: This should use the present tense:
```suggestion
| `setModuleMeta` | When compiling ESM to CJS, Babel defines a non-enumerable, non-witable, non-configurable `__esModule` property on the `exports` object. When this option is enabled, that property is set using a simple assignment | `loose` in `transform-modules-commonjs` |
```

> | `copyReexports` | When re-exporting an imported value, assume that it's value doesn't change and set it with a simple assignment | `loose` in `transform-modules-commonjs`, `transform-modules-umd`, `transform-modules-amd` _(not documented)_ | | |

Reviewer 4: Nit: `constantReexports`

```javascript 
// input code

foo?.bar;
```

```javascript
// output code

(() => {
  const _tmp = foo;
  if (_tmp == null && _tmp !== null && _tmp !== undefined) {
    throw new Error(`
      Invalid assumption - noDocumentAll.
      
      "noDocumentAll: true" implies that ?. and ?? are never
      used with the special document.all object.
      This assumption has been violated, so you should
      disable it in your config.
    `);
  }
  return _tmp;
})?.bar
```

<details>
<summary>Config</summary>

```jsonc
// babel.config.json
{
  "assumptions": {
    "noDocumentAll": true
  },
  "env": {
    "test": {
      "plugins": ["@babel/validate-assumptions"]
    }
  }
}
```

</details>

# Open questions

- Should we try to pass to the presets at least a _partial_ `assumptions` object? Currently `@babel/preset-env` relies on `loose` to enable/disable the `typeof-symbol` plugin.

> - Should we validate the list of `assumptions` in `@babel/core`, and disallow unknown ones? This would make it impossible for third-party plugins to introduce their own assumptions, but it also means that it's easier for us to introduce new assumptions without risking ecosystem incompatibilities.

Author: https://github.com/babel/rfcs/pull/5#issuecomment-641547262 (@existentialism)
> 2. My gut/initial reaction is that we should validate the list of `assumptions`

---

https://github.com/babel/rfcs/pull/5#issuecomment-643668256 (@nicolo-ribaudo)

> I initially wasn't 100% on board with it because it means that if we add support for a new assumptions in a plugin, then the user would need to update `@babel/core` in order to use it (otherwise Babel would throw an "unknown assumption" error).
However, I think that the benefits of validating outweigh this downside :+1:

---

I was going to write in the RFC text that `@babel/core` will throw on unknown assumptions, but I realized that if we add a new assumptions in `@babel/core@7.15`, presets cannot use it because it would be a breaking change when used with `@babel/core@7.14`.

Note that this problem will likely never affect our own presets (since we won't enable assumptions in our own presets), but it could affect framework-specific ones.

We have a few options:
1. Decide that it's ok, presets can check Babel's version and conditionally enable an assumption.
1. Only throw for unknown assumption in config files, not in presets.
1. Throw in config files, warn in presets.
1. Always warn.
1. Don't validate the assumptions list.

In any case, we shouldn't pass unknown assumptions down to plugins.

Reviewer 2: Agreed. Note that `@babel/preset-env` also validate unknown options. If a preset depends on `@babel/preset-env`, every time when a new option is introduced, they have to bump dependencies, too. I think it applies to `assumptions` of `@babel/core`, too.

| `iterableIsArray` | When using an iterable (in array destructuring, for-of or with spreads), assume that the iterable object is an `Array` | `loose` in `transform-destructuring` and `transform-spread`, `assumeArray` in `transform-for-of` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBCELwQ2gOmQIwLoG4BQpLwB4A0EyiAnmrFNkA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) (it doesn't support `assumeArray`) | 1 |
| `arrayLikeIsIterable` | Allow array-like objects to be used where an iterable is expected. This can be useful, for example, to iterate DOM collections in older browsers | `allowArrayLike` in `transform-destructuring`, `transform-spread` and `transform-for-of` | | 1 |
| `skipForOfIterationClosing` | When using `for-of` with an iterator, it should always be closed with `.return()` and with `.throw()` in case of an error. This option allows skipping those methods | `loose` in `transform-for-of` | | 2 |

> | `arrayIndexedIteration` | When using `for-of` with an array, ignore the iterator protocol and use direct accss to its element like a `for` loop would do | `loose` in `transform-for-of` | | 2 |

Author: While implementing this assumption, I realized that it's not something we currently do.

We fallback to accessing array elements not based on the loose option, but if the current engine doesn't have `Symbol` support.
I think we can delete this assumption from this list.

| `superAsFunctionCall` | When this option is enabled, `super(arg1)` will be transpiled to `BaseClass.call(this, arg1)`. This means that it won't work with native classes or with built-ins, but only with compiled classes or ES5 constructors | `loose` in `transform-classes` | |
| `inheritAsObjectCreate` | Use the `inheritsLoose` helper instead of `inherits` for derived classes | `loose` in `transform-classes` | |
| `noDocumentAll` | When compiling the `??` and `?.` operators, assume that they are never used with `document.all` and thus `== null` is safe | `loose` in `proposal-optional-chaining` and `proposal-nullish-coalescing-operator` | [🔗](https://babeljs.io/repl#?browsers=chrome%2070&build=&builtIns=false&spec=false&loose=false&code_lz=IYfgdARg3AUMAEIT2kA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) | 4 |

> | `pureGetters` | When an expression that might invoke getters needs to be evaluated multiple times, Babel caches its value. When this option is enabled, the expression can be safely re-evaluated. For example, `a.b?.()` can be compiled to `a.b != null && a.b()` without caching `a.b` | `loose` in `proposal-optional-chaining` | [🔗](https://babeljs.io/repl#?browsers=chrome%2070&build=&builtIns=false&spec=false&loose=true&code_lz=IYOgRg_CAUCUQ&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env&prettier=false&targets=&version=7.10.1&externalPlugins=) |

Author: When discussing with @JLHwung and @jridgewell, we couldn't decide between `pureGetters` and `noGetterSideEffects`.
Do you have any opinion on this?

Reviewer 3: I personally prefer `pureGetters`.

| `pureGetters` | When an expression that might invoke getters needs to be evaluated multiple times, Babel caches its value. When this option is enabled, the expression can be safely re-evaluated. For example, `a.b?.()` can be compiled to `a.b != null && a.b()` without caching `a.b` | `loose` in `proposal-optional-chaining` and in `proposal-object-rest-spread` _(not documented)_ | [🔗](https://babeljs.io/repl#?browsers=chrome%2070&build=&builtIns=false&spec=false&loose=true&code_lz=IYOgRg_CAUCUQ&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env&prettier=false&targets=&version=7.10.1&externalPlugins=) |

> | `setModuleMeta` | When compiling ESM to CJS, Babel defines a non-enumerable, non-witable, non-configurable `__esModule` property on the `exports` object. When this option is enabled, that property is set using a simple assignment | `loose` in `transform-modules-commonjs` |

Author: I was implementing this option, and what do you think about renaming it to `noModuleObjectIteration`?

The difference when this assumption is set, is that `__esModule` is enumerable so it will appear when iterating over a module object.

Reviewer 3: I personally prefer `enumerableModuleMeta`.

| `constantSuper` | The `super` binding in classes can be changed using `setPrototypeOf`, so it's not possible to statically know it. With this option Babel can assume that the superclass is never changed at runtime. | `loose` in `transform-classes`, `proposal-class-properties`, `proposal-private-methods` and `proposal-decorators`. It should also be added to `transform-object-super`. | [🔗](https://babeljs.io/repl#?browsers=chrome%2040&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEBC0G8BQ1oFsCmAXAFgewBMAKASkWjBAwCcsiAiWAQnpIG5oBfJbpUSGAGFEKdNnzEyCClVoNBLdlx5I-4KNACC0DAA8sGAHYEY8ZKmoBXQ6RGpUAegfQAKgAkAkgGVoAGQ8AcgCiTKKoEJYADjQAdJi4hKRsoty8APIARgBWGMBYMRDYAArUeFhlAJ7RaQBmRJoxkaXlWFUYADTQgo3NldHsqoYYAO5apDFWNuxAA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env&prettier=false&targets=&version=7.10.1&externalPlugins=) |

> | `noCallClass` | Assume that classes are always instantiated with `new` and that the code never tries to call them as function. This lets Babel skipping the `this instanceof ThisClass` check (`_classCallCheck`). | `loose` in `transform-classes` | |

Author: While implementing the remaining class-related assumptions, I noticed that this RFC was missing this one. In loose mode Babel assumes that you never do something like this:
```js
class A {}

A();
```
so it skips injecting the `_classCallCheck` helper.

This assumptions is the opposite of `newableArrowFunctions`. I was initially going to call this `callableClasses`, but then realized that these are bad names: we are not assuming that classes are callable (because they should not be!), but we are assuming that the user never tries to call them.

How do you feel about the `noCallClass` name, and about renaming `newableArrowFunctions` to `noNewArrow`? (in the future we might also have `noNewMethod`).

Reviewer 5: `noClassCalls`?

Author: :+1: I initially used `noCallClass` so that it's the same as `noNewArrow` (no + action + target), but I already mistyped it as `noClassCall` different times.

> | `noClassCalls` | Assume that classes are always instantiated with `new` and that the code never tries to call them as function. This lets Babel skipping the `this instanceof ThisClass` check (`_classCallCheck`). | `loose` in `transform-classes` | |

Reviewer 3: ```suggestion
| `noClassCalls` | Assume that classes are always instantiated with `new` and that the code never tries to call them as a function. This lets Babel skip the `this instanceof ThisClass` check (`_classCallCheck`). | `loose` in `transform-classes` | |
```

| `enumerableModuleMeta` | When compiling ESM to CJS, Babel defines a non-enumerable, non-witable, non-configurable `__esModule` property on the `exports` object. When this option is enabled, that property is set using a simple assignment | `loose` in `transform-modules-commonjs` |
| `constantReexports` | When re-exporting an imported value, assume that it's value doesn't change and set it with a simple assignment | `loose` in `transform-modules-commonjs`, `transform-modules-umd`, `transform-modules-amd` _(not documented)_ | | |

> | `noNewArrows` | Assume that the code never tries to instantiatearrow function using `new`, so Babel can avoid injecting checks to prevent it | `spec` in `transform-arrow-functions` | | 4 |

Reviewer 5: ```suggestion
| `noNewArrows` | Assume that the code never tries to instantiate arrow function using `new`, so Babel can avoid injecting checks to prevent it | `spec` in `transform-arrow-functions` | | 4 |
```

what about concise methods, generators, other things that lack `.prototype`?

Author: This is definitely something that we should add (for example, I mentioned `noNewMethods` in https://github.com/babel/rfcs/pull/5#discussion_r567348593).
However, with this RFC I'm only trying to specify _existing_ assumptions of the `loose`/`spec` options, and checking for generators/methods/etc is not something that Babel does yet.

# Open questions

> - Should we try to pass to the presets at least a _partial_ `assumptions` object? Currently `@babel/preset-env` relies on `loose` to enable/disable the `typeof-symbol` plugin. **ANSWER:** No, presets can only _produce_ assumptions and not _consume_ them.

Reviewer 5: Why not? Must a preset provide an unconditional transform that does consume the assumptions, if it wishes to operate on them?

what if a preset is invalid when certain assumptions are made or not made?

Author: We had to choose: we either resolve assumptions _after_ resolving presets, or _before_ resolving presets.
In the first cast it's impossible to pass the resolved assumptions to presets (because we don't know them yet), in the second case it's impossible to set assumptions in presets (because they have already been resolved).

If a preset needs to change its behavior based on an assumption, I suggest moving that change down to the plugins that the preset enables.

Reviewer 5: meaning, the plugin would validate and throw an error at transform time, not at config parsing time?

Author: At plugin instantiation time:

```js
function myPlugin(api) {
  if (api.assumption("unsupportedAssumption")) {
    throw new Error();
  }

  return {
    visitor: { ... }
  };
}
```

this is in the same phase as "preset instantiation" (it happens after, but you cannot only trigger the preset instantiation phase but not the plugin instantiation phase using the Babel API).
